xmppserver/src/test/java/org/jivesoftware/openfire/archive/ArchiverTest.java

```diff
@@ -0,0 +1,343 @@
+        }
+        finally
+        {
+            // Teardown fixture.
+            archiver.stop();
+        }
+    }
+
+    /**
+     * The archiver is expected to wait for a certain amount of time, before
+     * writing data to the store (this is intended to allow the archiver to batch
+     * work). This test verifies that two sets of data that are written to the
+     * archiver within the 'grace period' are written in different batches.
+     */
+    @Test
+    public void testArchiverInDistinctBatchesWhenOutsideGrace() throws Exception
+    {
+        // Setup fixture.
+        final int maxWorkQueueSize = 100;
+        final Duration maxPurgeInterval = Duration.ofMillis( 5000 );
+        final Duration gracePeriod = Duration.ofMillis( 50 );
+        final DummyArchiver archiver = new DummyArchiver( "test", maxWorkQueueSize, maxPurgeInterval, gracePeriod );
+        final Thread thread = new Thread( archiver );
+
+        try
+        {
+            // Execute system under test.
+            thread.start();
+            archiver.archive( 1 );
+            Thread.sleep( gracePeriod.toMillis() * 2 ); // Timing is affected by thread scheduling, etc. Waiting for _exactly_ 'grace period' often is not long enough for this test.
+            archiver.archive( 2 );
+
+            // Verify result.
+            waitUntilArchivingIsDone( archiver, 2 );
+
+            assertFalse( archiver.store.isEmpty() );
+            assertEquals( 2, archiver.batches.size() );
+            final List<Integer> firstBatch = archiver.batches.get(0).data;
+            final List<Integer> secondBatch = archiver.batches.get(1).data;
+            assertEquals( 1, firstBatch.size() );
+            assertEquals( 1, (long) firstBatch.get( 0 ) );
+            assertEquals( 1, secondBatch.size() );
+            assertEquals( 2, (long) secondBatch.get( 0 ) );
+        }
+        finally
+        {
+            // Teardown fixture.
+            archiver.stop();
+        }
+    }
+
+    /**
+     * The amount of work that is batched by the archiver is limited. This test
+     * asserts that a batch of work will not exceed that limit, even if more
+     * work is available within the 'grace period'.
+     */
+    @Test
+    public void testArchiverNoMoreThanQueue() throws Exception
+    {
+        // Setup fixture.
+        final int maxWorkQueueSize = 1;
+        final Duration maxPurgeInterval = Duration.ofMillis( 5000 );
+        final Duration gracePeriod = Duration.ofMillis( 250 );
+        final DummyArchiver archiver = new DummyArchiver( "test", maxWorkQueueSize, maxPurgeInterval, gracePeriod );
+        final Thread thread = new Thread( archiver );
+
+        try
+        {
+            // Execute system under test.
+            thread.start();
+            final Instant start = Instant.now();
+            archiver.archive( 1 );
+            archiver.archive( 2 );
+            if ( Duration.between( start, Instant.now() ).compareTo( gracePeriod ) >= 0 )
```

Comment:
I'd be very tempted to put a Thread.sleep(1) in your dummy archiver to ensure the clock ticks at least once between each saved item to help avoid this test becoming flickery.

Reply:
I'm happy to do so, but I'm not sure how that would make it less flickery?

Reply:
I'm going to merge as-is. If tests turn out to be flickery, I do expect a surplus of I-told-you-so. :)